### PR TITLE
Handle automatically mapping IPv4 addresses to IPV6 for dual mode sockets, fixes #36192

### PR DESCRIPTION
**Pull request by stefansedich**

### CODE REVIEW

**Comment by marek-safar** [comment]
Please follow mono coding guideline

File: mcs/class/System/Test/System.Net.Sockets/SocketTest.cs [hunk 1]
```diff
@@ -4319,6 +4319,87 @@ public void SendAsyncFile ()
...
+
+		// Test case for bug #36192
+		[Test]
+		public void BindToIPV4EndPointUsingDualModelSocket() {
+			using(var s = new Socket(SocketType.Stream, ProtocolType.Tcp)) {
+				try {
+					var host = new IPEndPoint(IPAddress.Loopback, 0);
+					
+					s.Bind(host);
+				} catch(SocketException) {
```
Comment (marek-safar) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: This is not needed

File: mcs/class/System/System.Net.Sockets/Socket.cs [hunk 2]
```diff
@@ -1147,6 +1147,13 @@ public void Bind (EndPoint local_end)
 
 			if (local_end == null)
 				throw new ArgumentNullException("local_end");
+				
+#if NET_4_5
```
Comment (akoeplinger) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: we don't need these ifdefs anymore, the 4.5 api is now always enabled.

Comment (stefansedich) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: Good to know! This was modeled after the recent additions of the MapToIPV4/6 methods which were ifdefed. Will remove them!

File: mcs/class/System/Test/System.Net.Sockets/SocketTest.cs (hunk 1 above)
Comment (stefansedich) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: Any reason why? This was my first failing test case that demonstrated the issue so it seemed appropriate.

Comment (stefansedich) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: Sure makes sense it is covered below with the other tests now, will clean up and remove.

**Comment by stefansedich** [comment]
@marek-safar @akoeplinger updated as per comments, @marek-safar hopefully I followed all code coding guidelines, if I missed anything please let me know (this is the first time I have submitted a PR for mono so I am still trying to get my head around everything).

File: mcs/class/System/Test/System.Net.Sockets/SocketTest.cs [hunk 3]
```diff
@@ -4319,6 +4319,67 @@ public void SendAsyncFile ()
 			}
 		}
```
Comment (marek-safar) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: try/catch/finally is not needed in your tests.
catch - failure will be detected by nunit
finally - you have using which does same

File: mcs/class/System/System.Net.Sockets/Socket.cs [hunk 4]
```diff
@@ -1283,6 +1290,10 @@ public void Connect (EndPoint remoteEP)
 
 			if (is_listening)
 				throw new InvalidOperationException ();
+				
+			if(ep != null) {
```
Comment (marek-safar) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: missing space

File: mcs/class/System/System.Net.Sockets/Socket.cs [hunk 5]
```diff
@@ -1147,7 +1147,12 @@ public void Bind (EndPoint local_end)
 
 			if (local_end == null)
 				throw new ArgumentNullException("local_end");
-
+				
+			var ipEndPoint = local_end as IPEndPoint;
+			if(ipEndPoint != null) {
```
Comment (marek-safar) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: missing space

File: mcs/class/System/Test/System.Net.Sockets/SocketTest.cs (hunk 3 above)
Comment (stefansedich) [inline mcs/class/System/Test/System.Net.Sockets/SocketTest.cs]: Thanks, makes sense and cleans it up alot.

**Comment by ludovic-henry** [comment]
@stefansedich LGTM. As soon as you sign the CLA, I can merge your PR. Thank you!

**Comment by stefansedich** [comment]
Thanks! Just working with my company go get my CLA sorted, hopefully I will
have this resolved next week.

Thanks
Stefan

**Comment by stefansedich** [comment]
All sorted @ludovic-henry sorry about the delay!

**Comment by ludovic-henry** [comment]
Thank you @stefansedich for your contribution! And no worries for the delay

**Pull request merged**